### DISCUSSION TITLE
Issue when reading a zip file in Thunderbird

### DISCUSSION THREAD

**Discussion by bluesky335** [gildas-lormeau/zip.js#423] [Q&A] [answered]
I am writing a Thunderbird extension and encountered an error when reading a zip file attachment in the background page. 
The error message says ‘File format is not recognized’, but it works fine in Chrome.
![image](https://github.com/gildas-lormeau/zip.js/assets/13523898/ed120627-0809-4ed7-9689-d250cb859f48)

Thunder Bird version: 102.11.2

**Comment by gildas-lormeau** [comment]
That probably means the JS runtime does not support generators. A way to fix this issue would be to compile zip.js with babel. The build script already exists, see https://github.com/gildas-lormeau/zip.js/blob/master/rollup-es5.config.js. You should just have to clone the project, run the `npm install` commands (cf. the top comment in `rollup-es5.config.js`) and `npx rollup -c rollup-es5.config.js` in order to build the files in the `dist` folder and the `index.cjs` file.

**Reply by bluesky335** [reply]
Thanks for your reply.
I got an error when run `npx rollup -c rollup-es5.config.js`

![image](https://github.com/gildas-lormeau/zip.js/assets/13523898/42e90f2a-4ebe-4910-b736-732f704ab8c4)

**Reply by gildas-lormeau** [reply]
Thank you for the feedback,  I just fixed the error in the script.

**Reply by bluesky335** [reply]
After updating the code, it builds success. but still has the problem "File format is not recognized".

**Reply by gildas-lormeau** [reply]
I don't think this error is related to the issue. Could you share the code causing the error and show the stack trace please?

**Reply by bluesky335** [reply]
This is a demo of thunder bird extension that includes zip.js.

[thunder-bird-extension.zip](https://github.com/gildas-lormeau/zip.js/files/11828349/thunder-bird-extension.zip)

- Download this file and unzip it.
- Open thunder bird -> Settings -> Add-ons and Themes -> Debug Add-ons -> Temporary Extensions -> Load Temporary Add-on. 
- Select "manifest.json" and open it. 
- Click the "Inspect" button to open the debug window.
- Then open a mail that has zip attachment, you can see the error in debug window.

<img width="1337" alt="image" src="https://github.com/gildas-lormeau/zip.js/assets/13523898/cbab4da3-39de-47a3-8f1c-b1ebc10189bd">

**Reply by gildas-lormeau** [reply]
Thank you, it looks like there is some inconsistency between the size of the file being uncompressed and the structure of the zip file. Are you able to open this zip file with this demo? https://gildas-lormeau.github.io/zip-manager/ (click on the "Import zip" button). If it does not work then the only solution is to send me the zip file you're trying to read.

**Reply by bluesky335** [reply]
Works fine with demo https://gildas-lormeau.github.io/zip-manager/.  

The zip file is just a normal zip file packaged by windows 11 Explorer.  Select files and right click them then choose “Compress to zip file”

![image](https://github.com/gildas-lormeau/zip.js/assets/13523898/d84667b2-8b7c-426f-9b27-f3b55bcd5200)

And this repository (https://github.com/Stuk/jszip) works fine with Thunderbird extension too.

**Reply by gildas-lormeau** [reply]
It's weird because that would mean it's a bug in Babel and I don't think it's the case. Could you please share the code where you create a `ZipReader` instance and show how you call it? I'm interested in the parameter of the constructor. I would like to know how you retrieve the zip file as binary data, and how you pass the data to the constructor.

**Reply by bluesky335** [reply]
You can download demo code [here ](https://github.com/gildas-lormeau/zip.js/discussions/423#discussioncomment-6247759) 

This is the code. You can find it at "/background/attachment-tool.js"
> The document of ` browser.messageDisplay`  and  `browser.messages` API
> [messageDisplay — WebExtension API Documentation for Thunderbird BetaManifest v2 documentation](https://webextension-api.thunderbird.net/en/latest/messageDisplay.html#getdisplayedmessage-tabid)
> [messages — WebExtension API Documentation for Thunderbird BetaManifest v2 documentation](https://webextension-api.thunderbird.net/en/latest/messages.html#listattachments-messageid)



``` javascript
/**
 * 
 * @param {message from DisplayScripts} message 
 * @param {sender} sender 
 * @returns 
 */
export default async function checkAttachment(message, sender) {
  // get current displayed message's info
  const messageHeader = await browser.messageDisplay.getDisplayedMessage(sender.tab.id);
  // get attachment info list of this mesage
  const attachmentList = await browser.messages.listAttachments(messageHeader.id)
  if (attachmentList.length === 0) {
    console.log("no attachment");
    return
  }
  // get attachment File object 
  let attachmentFile = await browser.messages.getAttachmentFile(messageHeader.id, attachmentList[0].partName)

  const zipFileReader = new zip.BlobReader(attachmentFile);
  const zipReader = new zip.ZipReader(zipFileReader);
  const entries = await zipReader.getEntries();
  console.log("check-attachment", attachmentList, entries);
}

```

**Reply by gildas-lormeau** [reply]
I tried your demo in Thunderbird/102.12.0. It works fine for me, I attached the zip file of your extension in a (draft) mail and I don't see any errors when displaying the mail in Thunderbird. The following log is displayed in the background page of the extension:
```
check-attachment Array [ {…} ] Array(17) [ {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, … ] attachment-tool.js:17:11
```

**Reply by bluesky335** [reply]
I tested the zip file of extension demo too, it works fine. But maybe you should try with a file that is larger than 1MB.
There is the file which I used.
- Two images of Genshin Impact: [genshin-images.zip](https://github.com/gildas-lormeau/zip.js/files/11844136/genshin-images.zip)
- Two Chromium license file: [LICENSES.chromium.zip](https://github.com/gildas-lormeau/zip.js/files/11844109/LICENSES.chromium.zip)

**Reply by gildas-lormeau** [reply, marked as answer]
Thank you, I was able to reproduce the issue. I've just published the version 2.7.17 which should fix it. BTW, you don't need to use the ES5 version.

**Comment by gildas-lormeau** [comment]
I'm moving the issue in the Discussions tab
